cpp/src/random_projection/rproj.cuh

```diff
@@ -85,7 +87,7 @@ void sparse_random_matrix(const raft::handle_t& h,
     size_t indices_idx = 0;
     size_t indptr_idx  = 0;
 
-    for (size_t i = 0; i < params.n_components; i++) {
+    for (int i = 0; i < params.n_components; i++) {
```

Comment:
Same story  for `n_components`. I believe `n_components` has no business being an `int`, so we should probably fix the type there.

Reply:
This is exactly why cuDF defines a `size_type` and uses it for all sizes and indices. That way there is never any ambiguity.

Reply:
`n_components` can't be `size_t` because of code like this: (rproj_utils.cuh line 143)

```
if (params.n_components == -1) {
    params.n_components = johnson_lindenstrauss_min_dim(params.n_samples, params.eps);
  }
```

Reply:
Also, since n_components can be negative, this loop needs to use `int`. This is bad -- probably need a better way than -1 as a sentinel here.

Reply:
Ahahahaha love it: 

```
inline void check_parameters(paramsRPROJ& params)
{
  ASSERT(params.n_components > 0, "Parameter n_components: must be strictly positive");
  ...
}

inline void build_parameters(paramsRPROJ& params)
{
  if (params.n_components == -1) {
    params.n_components = johnson_lindenstrauss_min_dim(params.n_samples, params.eps);
  }
  ...
}
```

Reply:
And `-1` is explicitly passed in tests. It's a feature.

Reply:
Blegh. Seems like we either need to change how we structure our parameters or look at using `std::optional`. Either way, seems like it's out of scope for this PR.